From a7bfca7a0ceeb18d58af792ccf6688c4bc8a3a8b Mon Sep 17 00:00:00 2001
From: Paul Thomson <paulthomson@google.com>
Date: Fri, 17 Apr 2026 11:20:25 +0100
Subject: [PATCH 9/9] IDEA-383770 [java] Fix PsiNewExpressionImpl.multiResolve
 and PsiParameterImpl.getLambdaParameterType

PR#3371

GitOrigin-RevId: bc73aea1db37e536e8ca052204c7044568c62180

https://github.com/JetBrains/intellij-community/commit/3e87a8c19a506652d40de63bbea35d54d7f7327c
https://youtrack.jetbrains.com/issue/IDEA-383770
Bug: 469501865
---
 .../psi/impl/source/PsiParameterImpl.java     | 54 ++++++++++++-------
 .../tree/java/PsiNewExpressionImpl.java       |  4 +-
 2 files changed, 38 insertions(+), 20 deletions(-)

diff --git a/java/java-psi-impl/src/com/intellij/psi/impl/source/PsiParameterImpl.java b/java/java-psi-impl/src/com/intellij/psi/impl/source/PsiParameterImpl.java
index 8c02a544804b..b3cd57eb2306 100644
--- a/java/java-psi-impl/src/com/intellij/psi/impl/source/PsiParameterImpl.java
+++ b/java/java-psi-impl/src/com/intellij/psi/impl/source/PsiParameterImpl.java
@@ -27,6 +27,7 @@ import org.jetbrains.annotations.Nullable;
 
 import javax.swing.*;
 import java.util.Arrays;
+import java.util.List;
 
 public class PsiParameterImpl extends JavaStubPsiElement<PsiParameterStub> implements PsiParameter {
   private volatile PsiType myCachedType;
@@ -51,29 +52,44 @@ public class PsiParameterImpl extends JavaStubPsiElement<PsiParameterStub> imple
       if (parameterIndex > -1) {
         PsiLambdaExpression lambdaExpression = PsiTreeUtil.getParentOfType(param, PsiLambdaExpression.class);
         if (lambdaExpression != null) {
-          PsiType functionalInterfaceType = MethodCandidateInfo.ourOverloadGuard.doPreventingRecursion(param, false,
-                                                                                                              () -> LambdaUtil.getFunctionalInterfaceType(lambdaExpression, true));
-          if (functionalInterfaceType == null) {
-            Ref<PsiType> typeRef = Ref.create();
-            // Probably there are several candidates for the functional expression type but all of them have the same parameter type
-            LambdaUtil.processParentOverloads(lambdaExpression, t -> {
-              PsiType candidate = getTypeForFunctionalInterfaceType(lambdaExpression, t, parameterIndex);
-              PsiType prevType = typeRef.get();
-              if (prevType == null) {
-                typeRef.set(candidate);
+          PsiType type = MethodCandidateInfo.ourOverloadGuard.doPreventingRecursion(param, false, () -> {
+            PsiElement parent = lambdaExpression.getParent();
+            if (parent instanceof PsiExpressionList) {
+              PsiElement gParent = parent.getParent();
+              if (gParent instanceof PsiAnonymousClass) {
+                gParent = gParent.getParent();
               }
-              else {
-                if (!(prevType instanceof PsiLambdaParameterType) && !prevType.equals(candidate)) {
-                  typeRef.set(new PsiLambdaParameterType(param));
+              List<?> overloadStack = MethodCandidateInfo.ourOverloadGuard.currentStack();
+              if (gParent instanceof PsiCall &&
+                  (overloadStack.contains(gParent) || overloadStack.contains(lambdaExpression))
+              ) {
+                return null;
+              }
+            }
+
+            PsiType functionalInterfaceType = LambdaUtil.getFunctionalInterfaceType(lambdaExpression, true);
+            if (functionalInterfaceType == null) {
+              Ref<PsiType> typeRef = Ref.create();
+              // Probably there are several candidates for the functional expression type but all of them have the same parameter type
+              LambdaUtil.processParentOverloads(lambdaExpression, t -> {
+                PsiType candidate = getTypeForFunctionalInterfaceType(lambdaExpression, t, parameterIndex);
+                PsiType prevType = typeRef.get();
+                if (prevType == null) {
+                  typeRef.set(candidate);
                 }
+                else {
+                  if (!(prevType instanceof PsiLambdaParameterType) && !prevType.equals(candidate)) {
+                    typeRef.set(new PsiLambdaParameterType(param));
+                  }
+                }
+              });
+              if (!typeRef.isNull()) {
+                return typeRef.get();
               }
-            });
-            if (!typeRef.isNull()) {
-              return typeRef.get();
             }
-          }
-          PsiType lambdaParameterFromType = getTypeForFunctionalInterfaceType(lambdaExpression, functionalInterfaceType, parameterIndex);
-          if (lambdaParameterFromType != null) return lambdaParameterFromType;
+            return getTypeForFunctionalInterfaceType(lambdaExpression, functionalInterfaceType, parameterIndex);
+          });
+          if (type != null) return type;
         }
       }
     }
diff --git a/java/java-psi-impl/src/com/intellij/psi/impl/source/tree/java/PsiNewExpressionImpl.java b/java/java-psi-impl/src/com/intellij/psi/impl/source/tree/java/PsiNewExpressionImpl.java
index 348d55b01c3c..12d52c48827a 100644
--- a/java/java-psi-impl/src/com/intellij/psi/impl/source/tree/java/PsiNewExpressionImpl.java
+++ b/java/java-psi-impl/src/com/intellij/psi/impl/source/tree/java/PsiNewExpressionImpl.java
@@ -209,7 +209,9 @@ public class PsiNewExpressionImpl extends ExpressionPsiElement implements PsiNew
 
   @Override
   public JavaResolveResult @NotNull [] multiResolve(boolean incompleteCode) {
-    return (JavaResolveResult[])getConstructorFakeReference().multiResolve(incompleteCode);
+    ResolveResult[] results = getConstructorFakeReference().multiResolve(incompleteCode);
+    if (results instanceof JavaResolveResult[]) return (JavaResolveResult[])results;
+    return ContainerUtil.map2Array(results, JavaResolveResult.EMPTY_ARRAY, result -> (JavaResolveResult)result);
   }
 
   @Override
-- 
2.54.0.rc1.513.gad8abe7a5a-goog

